src/core/push-notifications/PushNotificationService.ts

```
@@ -70,51 +70,48 @@ export default class PushNotificationService {
   }
 
   private async savePushToken(pushToken: PushToken) {
-    pushToken.lastUpdated = now();
     return await this.storage.setObject<PushToken>(KEY_PUSH_TOKEN, pushToken);
   }
 
   private tokenNeedsRefreshing(pushToken: PushToken) {
     return isDateBefore(pushToken.lastUpdated, aWeekAgo());
   }
 
-  private async sendPushToken(pushToken: PushToken) {
-    await this.apiClient.updatePushToken(pushToken);
-    await this.savePushToken(pushToken);
-  }
-
-  async refreshPushToken() {
+  async subscribeForPushNotifications() {
     try {
-      const pushToken = await this.getSavedPushToken();
+      // Always look to our local storage first.
+      let notify_backend = false;
+      let pushToken = await this.getSavedPushToken();
+
       if (!pushToken) {
-        await this.initPushToken();
+        // We don't even have a push token yet - fetch one!
+        pushToken = await this.fetchProviderPushToken();
+        notify_backend = true;
+        Analytics.track(events.NOTIFICATION_ENABLED);
       } else if (this.tokenNeedsRefreshing(pushToken)) {
-        await this.sendPushToken(pushToken);
+        // We need to re-fetch because it's been a while and might have changed.
+        pushToken = await this.fetchProviderPushToken();
+        notify_backend = true;
+        Analytics.track(events.NOTIFICATION_REFRESHED);
       }
-    } catch (error) {
-      // Do nothing.
-    }
-  }
 
-  async initPushToken() {
-    const pushToken = await this.createPushToken();
-    if (pushToken) {
-      try {
-        await this.sendPushToken(pushToken);
-        Analytics.track(events.NOTIFICATION_ENABLED);
-      } catch (error) {
-        // Fail silently
+      if (notify_backend && pushToken) {
+        // Send to our backend first and save locally only once successful.
+        await this.apiClient.updatePushToken(pushToken);
+        await this.savePushToken(pushToken);
       }
+    } catch (error) {
+      // Fail silently
```

Probably want to do something here...

---

Fail Noisily @andyguest :)

---

I'm maintaining the old error handling code here 😁 , but note that the silent failure still triggers a retry on next app opening.  Not sure if we can do much better than that?  Don't think we want to interrupt users with a pop-up for example, but maybe you'd prefer we did?